felix/wireguard/bootstrap.go

```diff
@@ -0,0 +1,70 @@
+package wireguard
+
+import (
+	"context"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+
+	"github.com/projectcalico/calico/felix/netlinkshim"
+	"github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+	cerrors "github.com/projectcalico/calico/libcalico-go/lib/errors"
+	"github.com/projectcalico/calico/libcalico-go/lib/options"
+)
+
+// BootstrapHostConnectivity forces WireGuard peers with hostencryption enabled to communicate with this node unencrypted.
+// This ensures connectivity in scenarios where we have lost our WireGuard config, but will be sent WireGuard traffic
+// e.g. after a node restart, during felix startup, when we need to fetch config from Typha (calico/issues/5125)
+func BootstrapHostConnectivity(wgDeviceName string, nodeName string, calicoClient clientv3.Interface) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	thisNode, err := calicoClient.Nodes().Get(ctx, nodeName, options.GetOptions{})
+	cancel()
+	if err != nil {
+		log.WithError(err).Warn("could not fetch node config from datastore")
+		return err
+	}
+
+	storedPublicKey := thisNode.Status.WireguardPublicKey
+	kernelPublicKey := getPublicKey(wgDeviceName, netlinkshim.NewRealWireguard)
+
+	// if there is any config mismatch, wipe the datastore's publickey (forces peers to send unencrypted traffic)
+	if storedPublicKey != kernelPublicKey.String() {
+		thisNode.Status.WireguardPublicKey = ""
+
+		maxRetries := 3
+		for r := 0; r < maxRetries; r++ {
+			ctx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
+			_, err := calicoClient.Nodes().Update(ctx, thisNode, options.SetOptions{})
+			cancel()
+			if err != nil {
+				switch err.(type) {
+				case cerrors.ErrorResourceUpdateConflict:
+					log.Debug("conflict while clearing wireguard config, retrying update")
```

Comment:
This error is telling you that your "Compare and swap" operation failed.  To resolve it you need to re-read the Node resource and apply your change to the new revision.  I.e. the read and update should be in the same retry loop together.

Reply:
gotcha.
on a slightly related note, the next error handler below this one keeps failing with:
```
failed to clear wireguard config error=error with field Metadata.ResourceVersion = '' (field must be set for an Update request)
```
but despite that, the operation does resolve the issue.

The docsstrings for `GetOptions.ResourceVersion` say:
`if unset, then the result is returned from remote storage based on quorum-read flag`

So I thought this would all just work :thinking:

Reply:
no longer occurring after reworking this